### PR TITLE
fix(tianmu): fix alter table add/drop primary key.(#1071) 

### LINKED ISSUE #1071: bug: Adding a primary key constraint succeeds even if there is duplicate data
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/stoneatom/stonedb/blob/stonedb-5.7-dev/CONTRIBUTING.md).

### Please confirm if bug report does NOT exists already ?

- [X] I confirm there is no existing issue for this

### Describe the problem

```sql
mysql> create table ttt(id int,name varchar(5));
Query OK, 0 rows affected (0.02 sec)

mysql> insert into ttt values(1,'张三'),(2,'李四'),(1,'张三');
Query OK, 3 rows affected (0.01 sec)
Records: 3  Duplicates: 0  Warnings: 0

mysql> select * from ttt;
+------+--------+
| id   | name   |
+------+--------+
|    1 | 张三   |
|    2 | 李四   |
|    1 | 张三   |
+------+--------+
3 rows in set (0.00 sec)

mysql> alter table ttt add constraint pk primary key(id);
Query OK, 3 rows affected (0.04 sec)
Records: 3  Duplicates: 0  Warnings: 0

mysql> show create table ttt\G
*************************** 1. row ***************************
       Table: ttt
Create Table: CREATE TABLE `ttt` (
  `id` int(11) NOT NULL,
  `name` varchar(5) DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=TIANMU DEFAULT CHARSET=utf8mb4
1 row in set (0.00 sec)

mysql> select * from ttt;
+----+--------+
| id | name   |
+----+--------+
|  1 | 张三   |
|  2 | 李四   |
|  1 | 张三   |
+----+--------+
3 rows in set (0.01 sec)
```

### Expected behavior

```sql
mysql> alter table ttt add constraint pk primary key(id);
ERROR 1062 (23000): Duplicate entry '1' for key 'PRIMARY'
```

### How To Reproduce

_No response_

### Environment

./mysqld  Ver 5.7.36-StoneDB for Linux on x86_64 (build-)
build information as follow:
        Repository address: https://github.com/stoneatom/stonedb.git:HEAD
        Branch name: HEAD
        Last commit ID: b44a51ce8
        Last commit time: Date:   Fri Dec 2 16:11:06 2022 +0000
        Build time: Date: Mon Dec  5 06:12:44 UTC 2022

### Are you interested in submitting a PR to solve the problem?

- [X] Yes, I will!

### PR DESCRIPTION
**Pull request by DandreChen**
## Summary about this PR

Issue Number: close #1071 

[summary]
add mtr for alter table add/drop primary key
fix bug:in some cases, can insert same values by primary key after atler table add primary key.

## Tests Check List

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [ ] New Feature
- [x] Bug Fix
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

**Review by lujiashun — APPROVED** [review]
LGTM

File: storage/tianmu/core/engine.cpp [hunk 1]
```diff
@@ -1953,6 +1950,16 @@ void Engine::AddTableIndex(const std::string &table_path, TABLE *table, [[maybe_
   }
 }
 
+void Engine::DeleteTableIndex(const std::string &table_path, [[maybe_unused]] THD *thd) {
+  std::unique_lock<std::shared_mutex> index_guard(tables_keys_mutex);
+  if (index::TianmuTableIndex::FindIndexTable(table_path)) {
+    index::TianmuTableIndex::DropIndexTable(table_path);
```
Comment (RingsC) [inline storage/tianmu/core/engine.cpp:1969]: How to know `DropIndexTable` returns with succeed.  here, we can not know the status of `index::TianmuTableIndex::DropIndexTable`

File: storage/tianmu/core/engine.h [hunk 2]
```diff
@@ -124,8 +124,10 @@ class Engine final {
   common::TX_ID MaxXID() const { return max_xid; }
   void DeferRemove(const fs::path &file, int32_t cookie);
   void HandleDeferredJobs();
-  // support for primary key
+  // support for add primary key
```
Comment (RingsC) [inline storage/tianmu/core/engine.h]: `for adding`

File: storage/tianmu/core/engine.cpp [hunk 3]
```diff
@@ -769,17 +769,13 @@ void Engine::Rollback(THD *thd, bool all, bool force_error_message) {
...
-    index::TianmuTableIndex::DropIndexTable(table);
-    m_table_keys.erase(table);
-  }
   {
     std::unique_lock<std::shared_mutex> mem_guard(mem_table_mutex);
     TianmuMemTable::DropMemTable(table);
     mem_table_map.erase(table);
   }
 
+  DeleteTableIndex(table, thd);
```
Comment (RingsC) [inline storage/tianmu/core/engine.cpp]: Here, we should do check whether operation is ok or not. if not, it should issue an error message.

**Review by RingsC — CHANGES_REQUESTED** [review]

File: storage/tianmu/core/engine.cpp [hunk 4]
```diff
@@ -1094,6 +1090,7 @@ void Engine::RenameTable([[maybe_unused]] Transaction *trans_, const std::string
   filter_cache.RemoveIf([id](const FilterCoordinate &c) { return c[0] == int(id); });
   system::RenameFile(tianmu_data_dir / (from + common::TIANMU_EXT), tianmu_data_dir / (to + common::TIANMU_EXT));
   RenameRdbTable(from, to);
+  DeleteTableIndex(from, thd);
```
Comment (RingsC) [inline storage/tianmu/core/engine.cpp]: we should add some fault-point tests. DBUG_EXECUTE_IF(xxxx). we should make the drop operation is robust operation.

**Review by RingsC — CHANGES_REQUESTED** [review]

File: storage/tianmu/core/engine.cpp (hunk 1 above)
Comment (DandreChen) [reply storage/tianmu/core/engine.cpp:1969]: fix it. add return value

File: storage/tianmu/core/engine.h (hunk 2 above)
Comment (DandreChen) [reply storage/tianmu/core/engine.h]: fixed it.

File: storage/tianmu/core/engine.cpp (hunk 4 above)
Comment (DandreChen) [reply storage/tianmu/core/engine.cpp]: added it

File: storage/tianmu/core/engine.cpp (hunk 3 above)
Comment (DandreChen) [reply storage/tianmu/core/engine.cpp]: added return value to check

**Review by RingsC — APPROVED** [review]
LGTM

**Pull request merged**